scanner/services/matcher.go

```diff
@@ -16,18 +23,83 @@ import (
 // matcherService represents a vulnerability matcher gRPC service.
 type matcherService struct {
 	v4.UnimplementedMatcherServer
+	// indexer is used to retrieve index reports.
+	indexer indexer.Indexer
+	// matcher is used to match vulnerabilities with index contents.
 	matcher matcher.Matcher
+	// disableEmptyContents allows the vulnerability matching API to reject requests with empty contents.
+	disableEmptyContents bool
 }
 
-// NewMatcherService creates a new vulnerability matcher gRPC service.
-func NewMatcherService(matcher matcher.Matcher) *matcherService {
+// NewMatcherService creates a new vulnerability matcher gRPC service, to enable
+// empty content in enrich requests, pass a nil indexer.
+func NewMatcherService(matcher matcher.Matcher, indexer indexer.Indexer) *matcherService {
 	return &matcherService{
-		matcher: matcher,
+		matcher:              matcher,
+		indexer:              indexer,
+		disableEmptyContents: indexer == nil,
 	}
 }
 
-func (s *matcherService) GetVulnerabilities(_ context.Context, _ *v4.GetVulnerabilitiesRequest) (*v4.VulnerabilityReport, error) {
-	return nil, status.Error(codes.Unimplemented, "method GetVulnerabilities not implemented")
+func (s *matcherService) GetVulnerabilities(ctx context.Context, req *v4.GetVulnerabilitiesRequest) (*v4.VulnerabilityReport, error) {
+	ctx = zlog.ContextWithValues(ctx, "component", "scanner/service/matcher")
+	if err := validators.ValidateGetVulnerabilitiesRequest(req); err != nil {
+		return nil, errox.InvalidArgs.CausedBy(err)
+	}
+	// Get an index report to enrich: either using the indexer, or provided in the request.
+	var ir *claircore.IndexReport
+	var err error
+	if req.GetContents() == nil {
+		if s.disableEmptyContents {
+			zlog.Debug(ctx).Msg("no contents, rejecting")
+			return nil, errox.InvalidArgs.New("empty contents is disabled")
+		}
+		zlog.Debug(ctx).Msg("no contents, retrieving")
+		ir, err = s.retrieveIndexReport(ctx, req.GetHashId())
+	} else {
+		zlog.Info(ctx).Msg("has contents, parsing")
+		ir, err = s.parseIndexReport(req.GetContents())
```

Comment:
The return from `parseIndexReport` looks to be a generic error w/ message `internal error` - in this particular case should that be a `bad request` instead?

Reply:
It should be an internal error because I expect invalid input to be blocked or sanitized in the validator.

Reply:
Ahh yes, I see that now. Excellent.